Add R4 support for export of additionalAttributes

tstrass opened the pull request:
### Summary
This adds support for export of `additionalAttributes`, which is a continuation of the work started in https://github.com/projecttacoma/synthea/pull/17.

### New Behavior
- Attributes specified in `additionalAttributes` in the module will be propagated to corresponding FHIR resource in the FHIR bundles exported from synthea.
- If the `additionalAttributes` contains any malformed JSON, it will effectively just be ignored for that state. An error message and a stack trace will print, but the program is allowed to continue, behaving as if no additional attributes were specified.
- This new behavior only exists for FHIR R4.

### Code Changes

- Changes are contained to the `FhirR4` exporter class.
- The implementation might seem a little odd. It's different from my initial approach, and it's explained somewhat in the code comments. Basically, what I first tried to do was parse the additional attributes to FHIR objects, and then just set the value for the corresponding property on the FHIR resource Java object. Ultimately I gave up on that approach, because as far as I could tell there was no simple way to do that parsing for any possible data type. Instead, I serialize the FHIR resource to JSON, set the additional attributes by just modifying the JSON directly, and then parse the resource back into a Java object. It essentially boils down to the fact that it's easy to parse whole FHIR resources, but not individual properties of those resources.

### Testing

- Try adding `additionalAttributes` to any state of your favorite module, running that module, and checking the exported bundles for the attributes.
- As an example, here's a separate branch I used for testing with EXM130: https://github.com/projecttacoma/synthea/compare/additional_attributes_export...projecttacoma:test_additional_attributes_export
  - If you use EXM130, you'll need to add [this ValueSet bundle](https://github.com/DBCG/connectathon/blob/master/fhir4/bundles/EXM130_FHIR4-7.2.000/EXM130_FHIR4-7.2.000-files/valuesets-EXM130_FHIR4-7.2.000-bundle.json) to /src/main/resources/terminology
- Use some malformed JSON, like putting an integer for a boolean value, and verify that you get the expected failure behavior described above.

tstrass committed:
Add R4 support for export of additionalAttributes

Support had already been additionalAttributes in the GMF.
This takes those attributes and exports them in FHIR R4.

okeefm commented on src/main/java/org/mitre/synthea/export/FhirR4.java:
```diff
@@ -768,7 +770,11 @@ private static BundleEntryComponent encounter(Person person, BundleEntryComponen
+    if (encounter.additionalAttributes != null) {
+      entry.setResource(setAdditionalAttributes(encounterResource, encounter.additionalAttributes));
+    }
+      return entry;
```
NIT: This should be un-indented 2 spaces
```suggestion
    return entry;
```

okeefm commented on src/main/java/org/mitre/synthea/export/FhirR4.java:
```diff
@@ -2883,6 +2912,42 @@ private static BundleEntryComponent newEntry(Bundle bundle, Resource resource,
+   * value will be overwritten.
+   * @param resource The resource to which the attributes apply
+   * @param additionalAttributes The attributes to apply
+   * @return A new Resource object with the additional attributes
```
```suggestion
   * @return A new Resource object with the additional attributes, or the original Resource object if the attributes cannot be successfully applied
```

okeefm commented on src/main/java/org/mitre/synthea/export/FhirR4.java:
```diff
@@ -2883,6 +2912,42 @@ private static BundleEntryComponent newEntry(Bundle bundle, Resource resource,
+    // because the FHIR library only exposes a method to parse a FHIR Resource, and the attributes
+    // can be of any FHIR type.
+    for (java.util.Map.Entry<String, JsonElement> e : additionalAttributes.entrySet()) {
+      jo.remove(e.getKey());
```
According to the Gson Java specs, the `add` method should remove the property if it exists, so this `remove` call is probably not required.

https://github.com/google/gson/blob/master/gson/src/main/java/com/google/gson/JsonObject.java#L57

okeefm reviewed (changes requested):
When testing this with the EXM130 branch, we were unable to see an error message or stacktrace when we put in well-formed JSON that isn't FHIR-compliant (see below). Should we have seen those messages with the JSON from below?

Also, it might be worth adding some tests for the FHIR R4 Exporter, to demonstrate the expected behaviors for: valid FHIR JSON, invalid FHIR JSON (that is valid against the base JSON spec), and totally invalid JSON.

```
"additional_attributes": {
  "foo": "15"
}
```

tstrass commented:
> When testing this with the EXM130 branch, we were unable to see an error message or stacktrace when we put in well-formed JSON that isn't FHIR-compliant (see below). Should we have seen those messages with the JSON from below?

Ah yeah the case where a property key just doesn't exist on the target object, it will silently be ignored. An example of when you'd get the error message is if a string is given as the value of a known boolean property. It would be nice if we printed an error message in the `"foo": "15"` "unknown property" case, but unfortunately with the way it's implemented, our error handling is more or less just a propogation of the error handling in [IParser.parseResource](https://hapifhir.io/hapi-fhir/apidocs/hapi-fhir-base/ca/uhn/fhir/parser/IParser.html#parseResource(java.lang.Class,java.lang.String)). In other words, since `IParser.parseResource` silently ignores unknown properties, so do we. I figured this was probably acceptable behavior for this feature, but if you disagree, I think I could implement it by checking each JSON key against each property name on the target object. However, that would entail a lot of reflection and would only give us unknown property handling at the top level of the `additionalAttributes` object (as in, not with nested properties).

tstrass commented:
>Also, it might be worth adding some tests for the FHIR R4 Exporter, to demonstrate the expected behaviors for: valid FHIR JSON, invalid FHIR JSON (that is valid against the base JSON spec), and totally invalid JSON.

Yeah good call. I will work on some tests.

okeefm commented:
>In other words, since IParser.parseResource silently ignores unknown properties, so do we.

I think ideally this would be a nice thing to throw an error on, but I don't think it rises to the level of a must-have for this PR. Maybe we make another JIRA ticket for better error handling to do later?

tstrass committed:
Remove redundant call in additional attribute parsing

Also fix an indentation and improve a comment.

This came out of code review

tstrass commented:
>Also, it might be worth adding some tests for the FHIR R4 Exporter, to demonstrate the expected behaviors for: valid FHIR JSON, invalid FHIR JSON (that is valid against the base JSON spec), and totally invalid JSON.

Thinking the test should be for valid FHIR JSON and invalid FHIR JSON, but totally invalid JSON will be caught in the GMF rather than export.

okeefm commented:
>Thinking the test should be for valid FHIR JSON and invalid FHIR JSON, but totally invalid JSON will be caught in the GMF rather than export.

That's fine, yeah; we can only add changes for valid/invalid FHIR JSON.

tstrass committed:
Add unit tests for additional attributes export

Also some checkstyle fixes. Also in order to allow the unit
test to call the additional attributes export method, I changed
the method from private to package-private.

tstrass commented:
>Also, it might be worth adding some tests for the FHIR R4 Exporter, to demonstrate the expected behaviors for: valid FHIR JSON, invalid FHIR JSON (that is valid against the base JSON spec), and totally invalid JSON.

I added tests for valid FHIR JSON and invalid FHIR JSON, but totally invalid JSON will be caught in the GMF rather than export.

tstrass committed:
Fix various checkstyle issues in additional attributes stuff

No functional changes.

okeefm reviewed (approved):
Looks good. Thanks for adding those tests!
